backend/src/interactions/updateUnconfirmedContribution/UpdateUnconfirmedContribution.context.ts

```diff
@@ -23,36 +26,39 @@ export class UpdateUnconfirmedContributionContext {
    */
   public constructor(
     private id: number,
-    private input: ContributionArgs | AdminUpdateContributionArgs,
+    private input: ContributionArgs | AdminUpdateContributionArgs | ContributionMessageArgs,
     private context: Context,
   ) {
     if (!context.role || !context.user) {
       throw new LogError("context didn't contain role or user")
     }
   }
 
-  public async run(): Promise<{
+  public async run(
+    transactionEntityManager?: EntityManager,
+    relations?: FindOptionsRelations<Contribution>,
+  ): Promise<{
     contribution: Contribution
-    contributionMessage: ContributionMessage
+    contributionMessage: ContributionMessage | undefined
     availableCreationSums: Decimal[]
     createdByUserChangedByModerator: boolean
+    contributionChanged: boolean
   }> {
     let createdByUserChangedByModerator = false
     if (!this.context.role || !this.context.user) {
       throw new LogError("context didn't contain role or user")
     }
-    const contributionToUpdate = await Contribution.findOne({
-      where: { id: this.id },
-    })
+    const options: FindOneOptions<Contribution> = { where: { id: this.id }, relations }
+    let contributionToUpdate: Contribution | null
+    if (transactionEntityManager) {
+      contributionToUpdate = await transactionEntityManager.findOne(Contribution, options)
```

Comment:
a read-access will not need a transaction-context unless the isolation-level is configured to `dirty-read` or `uncommitted-read` to get possibly changed and uncommited data in your result-set.
But as i know do we not use such a low isolation-level to be sure to get only commited and consistent data in a resultset, right?

Reply:
From Typeorm Doc: 

> The most important restriction when working in a transaction is to ALWAYS use the provided instance of entity manager - transactionalEntityManager in this example. DO NOT USE GLOBAL ENTITY MANAGER. All operations MUST be executed using the provided transactional entity manager.

[typeorm-transactions](https://orkhan.gitbook.io/typeorm/docs/transactions#creating-and-using-transactions)

The previous code used a transaction with REPEATABLE READ for adding (admin) contribution messages so I added this code path even if this is not longer really necessary because I combine all db read queries into one. But before changing this I must check this more detailed in a future refactoring.

Reply:
It has to be said that, strictly speaking, the read accesses in the previous code were executed via a different database connection outside the transaction. But I suspect that this was done due to unawareness and not intentionally.

Reply:
The query runner create a new connection to mysql so for performance reason I'd like to not use it all over the place.